### ISSUE TITLE
Silence class lifecycle hooks

### ISSUE THREAD

**Issue by Kl0tl** [acdlite/recompose#219]
**TLDR;** Class inheritance is handy to transparently enhance components, preserving instance and static methods. Could be easier and safer to use if React lifecycle methods were silenced automatically when calling `toClass` based on a flag.

---

Let’s say I want to compose forms. I create a base component to style the browser `input` tag with my favorite CSS abstraction (inline styles, CSS Modules, whatever):

```js
class Input extends Component {
  focus() {
    this.refs.input.focus();
  }

  blur() {
    this.refs.input.blur();
  }

  render() {
    return <input ref="input" { ...this.props } />;
  }
}
```

There are common properties I’d like many components to share, so I create some higher order components:

```js
const required = Input =>
  class extends Component {
    render() {
      return <Input required { ...this.props } />;
    }
  };

// …

const RequiredInput = required(Input);
```

The latter will hide instance and any static methods. There’s three ways around it.
  + Exposing the wrapped component on the wrapper
  + Copying non react properties on the wrapper
  + Extending the wrapped component

## Exposing the wrapped component on the wrapper

```diff
const required = Input =>
  class extends Component {
+   static Wrapped = Input;
    render() {
-    return <Input required { ...this.props } />;
+    return <Input ref="wrapped" required { ...this.props } />;
    }
  };
```

First solution sounds appealing because of its apparent simplicity but forces consumers to know wether our components are raw or enhanced. I can’t simply swap an `Input` component for a `RequiredInput` without updating callers of its instance methods. Instance and static properties must be handled separately. Also refs doesn’t work with functionals components so my higher order components must branch:

```diff
+ import { isClassComponent } from 'recompose';

const required = Input =>
-  class extends Component {
-   static Wrapped = Input;
-   render() {
-     return <Input ref="wrapped" required { ...this.props } />;
-   }
- };
+ isClassComponent(Input)
+    ? class extends Component {
+         static Wrapped = Input;
+         render() {
+           return <Input ref="wrapped" required { ...this.props } />;
+         }
+       }
+    : Object.assign(props => <Input required { ...props } />, { Wrapped: Input });
```

[React Redux](https://github.com/reactjs/react-redux) does this, minus the dependency. Doesn’t bother with a ref unless explicitly told so with a `withRef` option. 

## Copying non react properties on the wrapper

```diff
+ import hoistNonReactMethods from 'hoist-non-react-methods';

const required = Input =>
-  class extends Component {
+  hoistNonReactMethods(class extends Component {
    render() {
-    return <Input required { ...this.props } />;
+    return <Input ref="child" required { ...this.props } />;
    }
-  };
+  }, Component);
```

Second solution works but requires another dependency (https://github.com/elado/hoist-non-react-methods) and a ref. Refs still don’t work with functionals components, requiring another dependency (https://github.com/mridgway/hoist-non-react-statics):

```diff
+ import { isClassComponent } from 'recompose';
import hoistNonReactMethods from 'hoist-non-react-methods';
+ import hoistNonReactStatic from 'hoist-non-react-statics';

const required = Input =>
- hoistNonReactMethods(class extends Component {
-   render() {
-     return <Input ref="child" required { ...this.props } />;
-   }
- }, Component);
+ isClassComponent(Input)
+   ? hoistNonReactMethods(class extends Component {
+        render() {
+          return <Input ref="child" required { ...this.props } />;
+        }
+      }, Component)
+   : hoistNonReactStatics(props => <Input required { ...props } />, Input);
```

I don’t know anyone using this solution, `hoist-non-react-methods` [doesn’t seem to have any dependents](https://www.npmjs.com/browse/depended/hoist-non-react-methods).

## Extending the wrapped component

Third solution, while seeming absurd considering React best practices, allows complete transparency with minimal effort:

```diff
const required = Input =>
  class extends Input {
+    static defaultProps = {
+      required: true,
+      ...Input.defaultProps || {}
+    };
    render() {
-     return <Input required { ...this.props } />;
+    return super.render();
    }
  };
```

Functional components are trivially supported:

```diff
import { toClass } from 'recompose';

const required = Input =>
-  class extends Input {
+  class extends toClass(Input) {
    static defaultProps = {
      required: true,
      ...Input.defaultProps || {}
    };
    render() {
      return super.render();
    }
  };
```

[React CSS Modules](https://github.com/gajus/react-css-modules) does this, minus the dependency. Branches on functional components to return functions.

However, with this approach props can be overridden by consumers and more importantly can’t depend on other props or context :( Let’s render the wrapped component manually instead of this weird looking super call:

```diff
import { toClass } from 'recompose';

const required = Input =>
  class extends toClass(Input) {
-   static defaultProps = {
-     required: true,
-     ...Input.defaultProps || {}
-   };
    render() {
-    return super.render();
+    return <Input { ...this.props } required />;
    }
  };
```

Everything looks fine: our props can’t be overridden, we can compute them based on context or other props if needed and lifecycle hooks are called twice. Wait.

By extending the wrapped component we inherit all its static and instance methods, including lifecycle hooks which are then called once for the wrapper component and once for the wrapped component. Fixing this is easy but tedious:

```diff
import { toClass } from 'recompose';

const required = Input =>
  class extends toClass(Input) {
+   componentWillMount() {}
+   componentDidMount() {}
+   componentWillReceiveProps() {}
+   componentWillUpdate() {}
+   componentWillUnmount() {}
    render() {
      return <Input { ...this.props } required />;
    }
  };
```

Would be great to silence React lifecycle hooks automatically inside `toClass`:

```diff
- const toClass = baseComponent => {
+ const toClass = (baseComponent, { lifecycle = true } = {}) => {
  if (isClassComponent(baseComponent)) {
-   return baseComponent
+   return lifecycle ? baseComponent : class extends baseComponent {
+       componentWillMount() {}
+       componentDidMount() {}
+       componentWillReceiveProps() {}
+       componentWillUpdate() {}
+       componentWillUnmount() {}
+   }
  }

// …
```

---

What do you think ? Is there a more idiomatic way to achieve this with `recompose` ?

**Comment by andrei-cocorean** [comment]
Why do you want to expose the `focus` and `blur` methods on your wrapper? Do you want to toggle the focus from the input's parent?

**Comment by Kl0tl** [comment]
@andrei-cocorean Yes. But `focus` and `blur` are irrelevant here, what matters is that some times instance methods are the easiest way to interface with stateful components.

**Comment by andrei-cocorean** [comment]
> some times instance methods are the easiest way to interface with stateful components

Using instance methods together with inheritance and wrapped components doesn't work as one would expect though. For example the following doesn't work:

```js
class Input extends Component {
  focus() {
    this.refs.input.focus();
  }
  blur() {
    this.refs.input.blur();
  }
  render() {
    return <input ref="input" { ...this.props } />;
  }
}

const required = Input =>
  class extends toClass(Input) {
    componentWillMount() {}
    componentDidMount() {}
    componentWillReceiveProps() {}
    componentWillUpdate() {}
    componentWillUnmount() {}
    render() {
      return <Input { ...this.props } required />;
    }
  };

const RequiredInput = required(Input)

const input = ReactDOM.render(<RequiredInput />, document.getElementById('root'))
input.focus()  // <-- the `focus` method on RequiredInput doesn't delegate to the rendered Input
```

You need to re-implement the methods in the subclass to fix it.

Just to the point of a more idiomatic way of overriding lifecycle methods with recompose, you can use function composition:

```js
import {compose, toClass} from 'recompose'

function silenceLifecycle (cls) {
  return class extends cls {
    componentWillMount() {}
    componentDidMount() {}
    componentWillReceiveProps() {}
    componentWillUpdate() {}
    componentWillUnmount() {}
  }
}

const required = Input =>
  class extends toClass(Input) {
    render() {
      return <Input { ...this.props } required />
    }
  }

const RequiredInput = compose(silenceLifecycle, required, toClass)(Input)
```

**Comment by Kl0tl** [comment]
Instance methods referring to refs fail, you’re right. The only way to prevent this would be to mutate props inside render and call super:

```js
const required = Input =>
  class extends Input {
    render() {
      this.props.required = true;
      return super.render();
    }
  }
```

But those "dynamic" props would not show on `React.createElement` output. I guess there’s only three reliable uses of class inheritance for transparent higher order components:
  + Decorate React static properties (`childContextTypes`, `contextTypes`, `defaultProps`, `propTypes` and `displayName`)
  + Hook into React lifecycle methods
  + Transform the output of `super.render`

A combinator to silence lifecycle hooks isn’t needed in those cases.
